### PR TITLE
198 inventory service

### PR DESCRIPTION
**Pull request by heshamMassoud**
#### Summary
Mainly the change is about [only fetching the channels that are needed](https://github.com/commercetools/commercetools-sync-java/pull/473/files#diff-87e0a1f00ebdcae0ead2786d7c69a993L53) during the reference resolution of Channels on inventories during inventory sync.

Other changes:
- Refactoring and making services generic with two generic abstract classes `BaseService` and `KeyBasedService` for services for resources that have keys.
- Refactor inventory sync implementation so that it is similar to the other sync modules implementation. 
- Address those issues: #198, #150, #397

#### TODOs
- #403 issue description should be reformulated as it will need more changes too big for the current PR. #235 for being a possible duplicate.
- Release notes.

### CODE REVIEW

File: src/main/java/com/commercetools/sync/services/impl/CartDiscountServiceImpl.java [hunk 1]
```diff
@@ -19,7 +19,7 @@
 import java.util.Set;
 import java.util.concurrent.CompletionStage;
 
-public class CartDiscountServiceImpl extends BaseService<CartDiscountDraft, CartDiscount, CartDiscountSyncOptions,
+public class CartDiscountServiceImpl extends KeyBasedService<CartDiscountDraft, CartDiscount, CartDiscountSyncOptions,
```
Comment (heshamMassoud) [inline src/main/java/com/commercetools/sync/services/impl/CartDiscountServiceImpl.java]: `BaseServiceWithKey` and `BaseService`

File: src/main/java/com/commercetools/sync/inventories/helpers/InventoryEntryIdentifier.java [hunk 2]
```diff
@@ -53,34 +55,46 @@ public static InventoryEntryIdentifier of(@Nonnull final InventoryEntry inventor
...
+    }
+
     @Override
-    public boolean equals(final Object obj) {
-        if (obj instanceof  InventoryEntryIdentifier) {
-            final InventoryEntryIdentifier that = (InventoryEntryIdentifier) obj;
-            return Objects.equals(that.getInventoryEntryChannelId(), inventoryEntryChannelId)
-                && Objects.equals(that.getInventoryEntrySku(), inventoryEntrySku);
+    public String toString() {
+        return "{" + "sku='" + inventoryEntrySku + '\'' + ", channelKey='" + inventoryEntryChannelKey + '\'' + '}';
```
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/inventories/helpers/InventoryEntryIdentifier.java]: I think the formatting of this toString method would be better with format method.

File: src/main/java/com/commercetools/sync/services/StateService.java [hunk 3]
```diff
@@ -9,12 +9,10 @@
 
 public interface StateService {
     /**
-     * Given a {@code key}, this method first checks if a cached map of state keys -&gt; ids is not empty.
-     * If not, it returns a completed future that contains an optional that contains what this key maps to in
-     * the cache. If the cache is empty, the method populates the cache with the mapping of all state keys to ids
-     * in the CTP project, by querying the CTP project for all states.
-     *
-     * <p>After that, the method returns a {@link CompletionStage}&lt;{@link Optional}&lt;{@link String}&gt;&gt;
+     * Given a {@code key}, this method first checks if a cached map of state keys -&gt; ids contains the key.
```
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/services/StateService.java]:
```suggestion
     * Given a {@code key}, this method first checks if a cached map of State keys -&gt; ids contains the key.
```
to make it consistent with the others 😄

File: src/main/java/com/commercetools/sync/services/impl/BaseService.java [hunk 4]
```diff
@@ -233,14 +238,16 @@
      * keys in the CTP project, defined in an injected {@link SphereClient}. A mapping of the key to the id
      * of the fetched resources is persisted in an in-memory map.
      *
-     * @param          keys set of state keys to fetch matching states by
+     * @param keys          set of state keys to fetch matching states by
```
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/services/impl/BaseService.java]: not sure about state keys..

File: src/main/java/com/commercetools/sync/inventories/helpers/InventoryEntryIdentifier.java (hunk 2 above)
Comment (heshamMassoud) [reply src/main/java/com/commercetools/sync/inventories/helpers/InventoryEntryIdentifier.java]: e422c69

File: src/main/java/com/commercetools/sync/services/impl/BaseService.java (hunk 4 above)
Comment (heshamMassoud) [reply src/main/java/com/commercetools/sync/services/impl/BaseService.java]: d61882f

File: src/main/java/com/commercetools/sync/services/impl/BaseService.java [hunk 5]
```diff
@@ -238,7 +238,7 @@
      * keys in the CTP project, defined in an injected {@link SphereClient}. A mapping of the key to the id
      * of the fetched resources is persisted in an in-memory map.
      *
-     * @param keys          set of state keys to fetch matching states by
+     * @param keys          set of keys to fetch matching states by
```
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/services/impl/BaseService.java]: `matching states by` again 😅

Comment (heshamMassoud) [reply src/main/java/com/commercetools/sync/services/impl/BaseService.java]: cae11b4

**Pull request merged**